File: src/agent/gateway/messages.py
import httpx
from ..config import CONFIG


async def _send_rich_text_message(
    chat_id: int, markdown: str, reply_to: int | None = None
):
    payload = {
        "chat_id": chat_id,
        "rich_message": {
            "markdown": markdown,
        },
    }
    if reply_to:
        payload["reply_parameters"] = {"message_id": reply_to}

    async with httpx.AsyncClient() as client:
        resp = await client.post(
            f"https://api.telegram.org/bot{CONFIG.TELEGRAM_BOT_TOKEN}/sendRichMessage",
            json=payload,
        )

    if resp.status_code != 200:
        return None

    return resp.json()

File: internal/gateway/messages.go
package gateway

import (
	"bytes"
	"context"
	"encoding/json"
	"fmt"
	"net/http"
)

// sendRichMessage posts to Telegram's custom sendRichMessage endpoint with a
// markdown payload. Port of messages.py. Returns (nil, nil) on a non-200 so the
// caller can fall back to a standard sendMessage.
func (g *Gateway) sendRichMessage(ctx context.Context, chatID int64, markdown string, replyTo int) (map[string]any, error) {
	payload := map[string]any{
		"chat_id": chatID,
		"rich_message": map[string]any{
			"markdown": markdown,
		},
	}
	if replyTo != 0 {
		payload["reply_parameters"] = map[string]any{"message_id": replyTo}
	}

	body, err := json.Marshal(payload)
	if err != nil {
		return nil, err
	}

	url := fmt.Sprintf("https://api.telegram.org/bot%s/sendRichMessage", g.cfg.TelegramBotToken)
	req, err := http.NewRequestWithContext(ctx, http.MethodPost, url, bytes.NewReader(body))
	if err != nil {
		return nil, err
	}
	req.Header.Set("Content-Type", "application/json")

	resp, err := g.richClient.Do(req)
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK {
		return nil, nil
	}

	var out map[string]any
	if err := json.NewDecoder(resp.Body).Decode(&out); err != nil {
		return nil, err
	}
	return out, nil
}
